src/webgpu/api/operation/rendering/basic.spec.ts

```diff
@@ -129,4 +137,191 @@ g.test('large_draw')
+    });
+    const colorAttachmentView = colorAttachment.createView();
+
+    const runPipeline = async (numInstances: number, numVertices: number) => {
+      const pipeline = t.device.createRenderPipeline({
+        vertex: {
+          module: t.device.createShaderModule({
+            code: `
+            struct Output {
+              [[builtin(position)]] position: vec4<f32>;
+              [[location(0)]] color: vec4<f32>;
+            };
+
+            fn selectValue(index: u32, maxIndex: u32) -> f32 {
+              let highOrMid = select(1.0, 0.5, index == maxIndex - 1u);
+              return select(0.0, highOrMid, index == 0u);
+            }
+
+            [[stage(vertex)]] fn main(
+                [[builtin(vertex_index)]] v: u32,
+                [[builtin(instance_index)]] i: u32)
+                -> Output {
+              let r = selectValue(v, ${numVertices}u);
+              let b = selectValue(i, ${numInstances}u);
+              let x = (r * 4.0 - 2.0) / 3.0;
+              let y = (b * 4.0 - 2.0) / -3.0;
+              return Output(vec4<f32>(x, y, 0.0, 1.0),
+                            vec4<f32>(r, 0.0, b, 1.0));
+            }
+            `,
+          }),
+          entryPoint: 'main',
+        },
+        fragment: {
+          module: t.device.createShaderModule({
+            code: `
+              [[stage(fragment)]] fn main([[location(0)]] color: vec4<f32>)
+                  -> [[location(0)]] vec4<f32> {
+                return color;
+              }
+              `,
+          }),
+          entryPoint: 'main',
+          targets: [{ format: 'rgba8unorm' }],
+        },
+        primitive: { topology: 'point-list' },
+      });
+
+      const encoder = t.device.createCommandEncoder();
+      const pass = encoder.beginRenderPass({
+        colorAttachments: [
+          {
+            view: colorAttachmentView,
+            storeOp: 'store',
+            loadValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
+          },
+        ],
+      });
+
+      pass.setPipeline(pipeline);
+      if (indexed) {
+        pass.setIndexBuffer(indexBuffer, 'uint32');
+      }
+
+      if (indirect) {
+        writeIndirectParams(numVertices, numInstances);
+        if (indexed) {
+          pass.drawIndexedIndirect(indirectBuffer, 0);
+        } else {
+          pass.drawIndirect(indirectBuffer, 0);
+        }
+      } else {
+        if (indexed) {
+          pass.drawIndexed(numVertices, numInstances);
+        } else {
+          pass.draw(numVertices, numInstances);
+        }
+      }
+      pass.endPass();
+      encoder.copyTextureToBuffer(
+        { texture: colorAttachment, mipLevel: 0, origin: { x: 0, y: 0, z: 0 } },
+        { buffer: dst, bytesPerRow: BYTES_PER_ROW },
+        { width: 3, height: 3, depthOrArrayLayers: 1 }
+      );
+      t.device.queue.submit([encoder.finish()]);
+
+      // Red should go 0, 127, 255 from left to right. Blue should to the same
+      // from top to bottom.
```

Comment:
Just a reminder for myself when I'm working on texture checking helpers soon.

```suggestion
      // from top to bottom.
      // TODO(crbug.com/dawn/868): See if this can be simplified with new texture checking helpers.
```